packages/runtime-core/__tests__/helpers/resolveAssets.spec.ts

```diff
@@ -0,0 +1,94 @@
+import {
+  mockWarn,
+  createApp,
+  nodeOps,
+  resolveComponent,
+  resolveDirective,
+  Component,
+  Directive
+} from '@vue/runtime-test'
+
+describe('resolveAssets', () => {
+  test('should work', () => {
+    const app = createApp()
+    const FooBar = () => null
+    const BarBaz = { mounted: () => null }
+
+    let component1: Component
+    let component2: Component
+    let component3: Component
+    let component4: Component
+    let directive1: Directive
+    let directive2: Directive
+    let directive3: Directive
+    let directive4: Directive
+
+    const Root = {
+      components: {
+        FooBar: FooBar
+      },
+      directives: {
+        BarBaz: BarBaz
+      },
+      setup() {
+        return () => {
+          component1 = resolveComponent('FooBar')!
+          directive1 = resolveDirective('BarBaz')!
+          // camelize
+          component2 = resolveComponent('Foo-bar')!
+          directive2 = resolveDirective('Bar-baz')!
+          // capitalize
+          component3 = resolveComponent('fooBar')!
+          directive3 = resolveDirective('barBaz')!
+          // camelize and capitalize
+          component4 = resolveComponent('foo-bar')!
+          directive4 = resolveDirective('bar-baz')!
+        }
+      }
+    }
+
+    const root = nodeOps.createElement('div')
+    app.mount(Root, root)
+    expect(component1!).toBe(FooBar)
```

Comment:
`component1` etc are of type `Component`, is this `!` needed here?

Reply:
it use to tell the `ts` that variable has been assigned